add lakectl local ang lakeFS<>Git integration docs (#7450)
treeverse/lakeFS: docs/howto/local-checkouts.md

---
title: Work with lakeFS data locally
description: How to work with lakeFS data locally
parent: How-To
---

# Working with lakeFS Data Locally

> lakeFS is a scalable data version control system designed to scale to billions of objects, and as such it has no notion of

Reviewer: Nitpick: I'd lead with the positive (i.e. instead of "lakeFS doesn't do X, however sometime it's needed", go with "with lakeFS you can do Y in addition to X!")

Reviewer: Maybe add another short paragraph to act as TL;DR?

> This document explains `lakectl local`, a command that lets you clone data from lakeFS to a local directory, keep it in sync, ...

"checking out" a branch. The larger the data, the less feasible it becomes to consume it from a single machine. However, 
there are scenarios where checking out lakeFS data becomes essential. 

{% include toc.html %} 

## Use cases 

### Local development of ML models

The development of machine learning models is a dynamic and iterative process, including experimentation with various data versions, 
transformations, algorithms, and hyperparameters. To optimize this iterative workflow, experiments must be conducted with speed, 
ease of tracking, and reproducibility in mind. Localizing the model data during development enhances the development process. 
It **accelerates the development process** by enabling interactive and offline development and reducing data access latency. 

> The local availability of data is required to **seamlessly integrate data version control systems and source control systems**

Reviewer: should be a different paragraph

like Git. This integration is vital for achieving model reproducibility, allowing for a more efficient and collaborative 
model development environment.

### Data Locality for Optimized GPU Utilization

> When executing deep learning tasks, GPUs, being valuable yet expensive resource, require efficient utilization. In the 

Reviewer: ```suggestion
Training Deep Learning models requires expensive GPUs. In the 
```

context of running such programs, the goal is to optimize GPU usage and prevent them from sitting idle. Many deep learning
tasks involve accessing images, and in some cases, the same images are accessed multiple times. Localizing the data can 

> eliminate redundant round trip times to access remote storage, resulting in cost savings.

Reviewer: consider adding an illustration

The _local_ command of lakeFS' CLI _lakectl_ enables working with lakeFS data locally.
It allows cloning lakeFS data into a directory on any machine, syncing local directories with remote lakeFS locations, 
and to [seamlessly integrate lakeFS with Git](#example-using-lakectl-local-in-tandem-with-git).

Here are the available _lakectl local_ commands: 

| Command                                  | What it does                                                                                                                                        | Notes                                                                                                                                                                                                                     |
|:-----------------------------------------|:----------------------------------------------------------------------------------------------------------------------------------------------------|:--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|

> | **init**                                 | Connects between a local directory and a lakeFS remote URI to enable data sync                                                                      | To undo a directory init, delete the .lakefs_ref.yaml file created in the initialized directory                                                                                                                           |

Reviewer: should each link to the cli reference for all the flags/options

| **pull**                                 | Fetches latest changes from a lakeFS remote location into a connected local directory                                                               |                                                                                                                                                                                                                           |
| **checkout**                             | Syncs a local directory with the state of a lakeFS ref                                                                                              |                                                                                                                                                                                                                           |

For additional information see the [CLI reference](../reference/cli.md#lakectl-local). 

> **Note:** The data size you work with locally should be reasonable for smooth operation on a local machine.

Reviewer: add a rule of thumb... is 500TB reasonable?

Reviewer: A number, even if a ballpark so that less technically capable readers will understand. Could be fixed later though.

![Git repo](../assets/img/lakectl-local/code_repo.png)

We also created a lakeFS repository and uploaded the _is_alpaca_ train [dataset](https://www.kaggle.com/datasets/sayedmahmoud/alpaca-dataset) 
by Kaggel into it:
![lakeFS repo](../assets/img/lakectl-local/lakefs-repo-with-train-dataset.png)

### Create an Isolated Environment for Experiments
{: .no_toc}

Our goal is to improve the model predictions. To meet our goal, we will experiment with editing the training dataset. 
We will run our experiments in isolation to not change anything until after we are certain the data is improved and ready.    

Let's create a new lakeFS branch called `experiment-1`. Our _is_alpaca_ dataset is accessible on that branch, 
and we will interact with the data from that branch only.
![experiment-1-branch](../assets/img/lakectl-local/experiment-branch.png)

On the code side, we will create a Git branch also called `experiment-1` to not pollute our main branch with a dataset
which is under tuning.

### Clone lakeFS Data into a Local Git Repository  
{: .no_toc}

Inspecting the `train.py` script, we can see that it expects an input on the `input` directory.

> <img src="../assets/img/lakectl-local/train_script_internals.png" alt="Expected input location" width="600"/>

Reviewer: Avoid screenshots of code, they can't be searched and are hard to modify later...

  (use "git restore <file>..." to discard changes in working directory)
	modified:   .gitignore
```

We can see that the `.gitignore` file changed, and that the files we cloned from lakeFS into the `input` directory are not 
tracked by git. This is intentional - remember that **lakeFS is the one managing the data**. But wait, what is this special
`input/.lakefs_ref.yaml` file that Git does track?  
```shell
is_alpaca % cat input/.lakefs_ref.yaml

src: lakefs://is-alpaca/experiment-1/dataset/train/s
at_head: 589f87704418c6bac80c5a6fc1b52c245af347b9ad1ea8d06597e4437fae4ca3
```
This file includes the lakeFS version of the **data** that the Git repository is currently pointing to.

Let's commit the changes to Git with:
```shell
git commit -m "added is_alpaca dataset" 
```
By committing to Git, we tie the current code version of the model to the dataset version in lakeFS as it appears in
`input/.lakefs_ref.yaml`.

### Experiment and Version Results  
{: .no_toc}

We ran the train script on the cloned input, and it generated a model. Now, let's use the model to predict whether **SpongeBob** is an Alpaca.

A reminder - this is how SpongeBob looks like - not like an Alpaca!

> <img src="../assets/img/lakectl-local/sponge-bob.png" alt="SpongeBob" width="300"/>

Reviewer: Either find an image that is CC0/public domain licensed (with proper attribution) or use another example :)

Reviewer: Also, Axolotls are cuter and are also not alpacas

Author: done, replaced with a cute axolotl :)
